scripts/leapp_preupgrade.py

```diff
@@ -0,0 +1,340 @@
+
+    if not wait:
+        return output, None
+
+    for line in iter(process.stdout.readline, b""):
+        line = line.decode("utf8")
+        output += line
+
+    # Call wait() to wait for the process to terminate so that we can
+    # get the return code.
+    process.wait()
+
+    return output, process.returncode
+
+
+def check_if_package_installed(pkg_name):
+    _, return_code = run_subprocess(["rpm", "-q", pkg_name])
+    return return_code == 0
+
+
+def setup_leapp(command, rhui_pkgs):
+    print("Installing leapp ...")
+    output, returncode = run_subprocess(command)
+    if returncode:
+        print(
+            "Installation of leapp failed with code '%s' and output: %s\n"
+            % (returncode, output)
+        )
+        raise ProcessError(
+            message="Installation of leapp failed with code '%s'." % returncode
+        )
+
+    print("Check installed rhui packages ...")
+    for pkg in rhui_pkgs:
+        if check_if_package_installed(pkg["src_pkg"]):
+            pkg["installed"] = True
+    return [pkg for pkg in rhui_pkgs if pkg.get("installed", False)]
+
+
+def should_use_no_rhsm_check(rhui_installed, command):
+    print("Checking if subscription manager and repositories are available ...")
+    rhsm_repo_check_fail = True
+    _, rhsm_installed_check = run_subprocess(["which", "subscription-manager"])
+    if rhsm_installed_check == 0:
+        rhsm_repo_check, _ = run_subprocess(
+            ["subscription-manager", "repos", "--list-enabled"]
+        )
+        rhsm_repo_check_fail = (
+            "This system has no repositories available through subscriptions."
+            in rhsm_repo_check
+            or "Repositories disabled by configuration." in rhsm_repo_check
+        )
+
+    if rhui_installed and not rhsm_repo_check_fail:
+        print("RHUI packages detected, adding --no-rhsm flag to preupgrade command")
+        command.append("--no-rhsm")
+        return True
+    return False
+
+
+def remove_previous_reports():
+    print("Removing previous preupgrade reports at /var/log/leapp/leapp-report.* ...")
+
+    json_report_path = "/var/log/leapp/leapp-report.json"
+    if os.path.exists(json_report_path):
+        os.remove(json_report_path)
+
+    text_report_path = "/var/log/leapp/leapp-report.txt"
+    if os.path.exists(text_report_path):
+        os.remove(text_report_path)
+
+
+def execute_preupgrade(command):
+    print("Executing preupgrade ...")
+    _, _ = run_subprocess(command)
+
+    # NOTE: we do not care about returncode because non-null always means actor error (or leapp error)
+    # if returncode:
+    #     print(
+    #         "The process leapp exited with code '%s' and output: %s\n"
+    #         % (returncode, output)
+    #     )
+    #     raise ProcessError(message="Leapp exited with code '%s'." % returncode)
+
+
+def parse_results(output):
+    print("Processing preupgrade results ...")
+
+    json_report_path = "/var/log/leapp/leapp-report.json"
```

Comment:
You could convert this to a global since you are calling it multiple times. Might even be easier to adapt the tests later as you will mock that global anyway.

Reply:
moved both report paths to the top as global constants